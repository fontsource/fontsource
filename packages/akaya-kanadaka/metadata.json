{
  "fontId": "akaya-kanadaka",
  "fontName": "Akaya Kanadaka",
  "subsets": ["kannada", "latin", "latin-ext"],
  "weights": ["400"],
  "styles": ["normal"],
  "defSubset": "latin",
  "variable": false,
  "lastModified": "2021-03-24",
  "version": "v4",
  "category": "display",
  "source": "https://fonts.google.com/",
  "license": "https://fonts.google.com/attribution",
  "type": "google"
}
