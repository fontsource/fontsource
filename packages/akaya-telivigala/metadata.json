{
  "fontId": "akaya-telivigala",
  "fontName": "Akaya Telivigala",
  "subsets": ["latin", "latin-ext", "telugu"],
  "weights": ["400"],
  "styles": ["normal"],
  "defSubset": "latin",
  "variable": false,
  "lastModified": "2021-02-18",
  "version": "v9",
  "category": "display",
  "source": "https://fonts.google.com/",
  "license": "https://fonts.google.com/attribution",
  "type": "google"
}
