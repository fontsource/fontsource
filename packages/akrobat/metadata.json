{"fontId":"akrobat","fontName":"Akrobat","subsets":["all"],"weights":[100,200,300,400,600,700,800,900],"styles":["normal"],"defSubset":"all","variable":false,"lastModified":"2022-02-18","version":"","category":"sans-serif","source":"https://www.fontfabric.com/fonts/akrobat/","license":"https://localfonts.eu/wp-content/uploads/2017/12/EULA-Free-Font-License-Ver.-2.0.pdf","type":"other"}
