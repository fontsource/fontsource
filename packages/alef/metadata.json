{
  "fontId": "alef",
  "fontName": "Alef",
  "subsets": ["hebrew", "latin"],
  "weights": ["400", "700"],
  "styles": ["normal"],
  "defSubset": "latin",
  "variable": false,
  "lastModified": "2020-09-02",
  "version": "v12",
  "category": "sans-serif",
  "source": "https://fonts.google.com/",
  "license": "https://fonts.google.com/attribution",
  "type": "google"
}
