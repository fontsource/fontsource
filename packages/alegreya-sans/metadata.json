{
  "fontId": "alegreya-sans",
  "fontName": "Alegreya Sans",
  "subsets": [
    "cyrillic",
    "cyrillic-ext",
    "greek",
    "greek-ext",
    "latin",
    "latin-ext",
    "vietnamese"
  ],
  "weights": ["100", "300", "400", "500", "700", "800", "900"],
  "styles": ["italic", "normal"],
  "defSubset": "latin",
  "variable": false,
  "lastModified": "2021-03-24",
  "version": "v14",
  "category": "sans-serif",
  "source": "https://fonts.google.com/",
  "license": "https://fonts.google.com/attribution",
  "type": "google"
}
