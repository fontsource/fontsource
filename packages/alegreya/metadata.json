{
  "fontId": "alegreya",
  "fontName": "Alegreya",
  "subsets": [
    "cyrillic",
    "cyrillic-ext",
    "greek",
    "greek-ext",
    "latin",
    "latin-ext",
    "vietnamese"
  ],
  "weights": ["400", "500", "600", "700", "800", "900"],
  "styles": ["italic", "normal"],
  "defSubset": "latin",
  "variable": {
    "ital": { "default": "0", "min": "0", "max": "1", "step": "1" },
    "wght": { "default": "400", "min": "400", "max": "900", "step": "1" }
  },
  "lastModified": "2021-02-12",
  "version": "v19",
  "category": "serif",
  "source": "https://fonts.google.com/",
  "license": "https://fonts.google.com/attribution",
  "type": "google"
}
