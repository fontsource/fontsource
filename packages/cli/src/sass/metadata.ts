import type { Axes, Metadata, UnicodeRange } from '../types';
import { findClosest } from '../utils';

const axesValue = (metadata: Metadata) => {
	let out = '(\n';
	for (const variable of Object.keys(metadata.variable)) {
		const axis = (metadata.variable as Axes)[variable];
		out += `  ${variable}: (
    default: ${axis.default},
    min: ${axis.min},
    max: ${axis.max},
    step: ${axis.step},
  ),
`;
	}
	out += ')';
	return out;
};

const unicodeValue = (unicode: UnicodeRange) => {
	if (Object.keys(unicode).length > 0) {
		let out = '(\n';

		for (const unicodeKey of Object.keys(unicode)) {
			const unicodeVal = unicode[unicodeKey];

			out += `  ${unicodeKey
				.replace('[', '')
				.replace(']', '')}: (${unicodeVal}),\n`;
		}

		out += ')';

		return out;
	}

	return 'null';
};

const defaultAxis = (metadata: Metadata) => {
	const axes = Object.keys(metadata.variable);
	if (axes.includes('wght')) return 'wght';
	if (axes.includes('full')) return 'full';
	return axes[0];
};

export const sassMetadata = (
	metadata: Metadata,
	unicode: UnicodeRange,
	isVariable: boolean,
) => {
	return `$metadata: (
  id: '${metadata.id}',
  family: '${metadata.family}',
  category: ${metadata.category},
  subsets: (${metadata.subsets.join(', ')}),
  weights: (${metadata.weights.join(', ')}),
  styles: (${metadata.styles.join(', ')}),
  axes: ${isVariable ? axesValue(metadata) : 'null'},
  defaults: (
    subset: ${metadata.defSubset},
    weight: ${findClosest(metadata.weights, 400)},
    style: normal,
    axis: ${isVariable ? defaultAxis(metadata) : 'null'},
  ),
  unicode: ${unicodeValue(unicode)}
) !default;`;
};
