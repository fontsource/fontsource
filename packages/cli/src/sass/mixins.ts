export const sassMixins = `@use 'sass:list';
@use 'sass:map';
@use 'sass:math';
@use 'sass:meta';
@use 'sass:string';

@use 'metadata';

$metadata: meta.module-variables(metadata) !default;
$directory: null !default;

$family: null !default;
$display: null !default;
$displayVar: null !default;
$formats: null !default;
$subsets: null !default;
$weights: null !default;
$styles: null !default;
$axes: null !default;

@mixin generator(
  $metadata: $metadata,
  $directory: $directory,
  $family: $family,
  $display: $display,
  $displayVar: $displayVar,
  $formats: $formats,
  $subsets: $subsets,
  $weights: $weights,
  $styles: $styles,
  $axes: $axes
) {
  $isVariable: map.get($metadata, axes) != null;

  $directory: if(
    $directory,
    $directory,
    '~@fontsource#{if($isVariable, '-variable', '')}/#{map.get($metadata, id)}/files'
  );

  $family: if($family, $family, map.get($metadata, family) + if($isVariable, ' Variable', ''));
  $display: if($display, $display, swap);
  $displayVar: if($displayVar != null, $displayVar, true);
  $formats: if(not $formats or $formats == all, if($isVariable, woff2, (woff2, woff)), $formats);
  $subsets: if(
    $subsets,
    if($subsets == all, map.get($metadata, subsets), $subsets),
    map.get($metadata, defaults, subset)
  );
  $weights: if(
    $weights,
    if($weights == all, map.get($metadata, weights), $weights),
    map.get($metadata, defaults, weight)
  );
  $styles: if(
    $styles,
    if($styles == all, map.get($metadata, styles), $styles),
    map.get($metadata, defaults, style)
  );
  $axes: if(
    $axes,
    if($axes == all, full, $axes),
    if($isVariable, if(map.has-key($metadata, axes, wght), wght, full), null)
  );

  @each $subset in $subsets {
    @each $unicodeSubset, $unicodeRange in map.get($metadata, unicode) {
      @if (
        ($subset == $unicodeSubset) or
          (
            // Is numeric subset
            ($subset == map.get($metadata, defaults, subset)) and not
              list.index(map.get($metadata, subsets), $unicodeSubset)
          )
      ) {
        @each $weight in if($axes, null, $weights) {
          @each $axis in $axes {
            @each $style in $styles {
              $variant: '#{map.get($metadata, id)}-#{if($unicodeSubset, $unicodeSubset, $subset)}-#{if($axis, $axis, $weight)}-#{$style}';

              $src: ();
              @each $format in $formats {
                $src: append(
                  $src,
                  url('#{$directory}/#{$variant}.#{$format}')
                    format('#{$format}#{if($axis, '-variations', '')}'),
                  comma
                );
              }

              @content ((
                metadata: $metadata,
                directory: $directory,
                family: $family,
                display: $display,
                displayVar: $displayVar,
                formats: $formats,
                subsets: $subsets,
                weights: $weights,
                styles: $styles,
                axes: $axes,

                variant: $variant,
                subset: $subset,
                unicodeSubset: $unicodeSubset,
                unicodeRange: $unicodeRange,
                weight: $weight,
                axis: $axis,
                style: $style,

                font-family: string.quote($family),
                font-style: if(
                  (($axis == full) or ($axis == slnt)) and map.has-key($metadata, axes, slnt),
                  oblique map.get($metadata, axes, slnt, min) + deg map.get($metadata, axes, slnt, max) + deg,
                  $style
                ),
                font-display: if($displayVar, var(--fontsource-display, $display), $display),
                font-weight: if(
                  (($axis == full) or ($axis == wght)) and map.has-key($metadata, axes, wght),
                  map.get($metadata, axes, wght, min) map.get($metadata, axes, wght, max),
                  $weight
                ),
                font-stretch: if(
                  (($axis == full) or ($axis == wdth)) and map.has-key($metadata, axes, wdth),
                  '#{map.get($metadata, axes, wdth, min)}% #{map.get($metadata, axes, wdth, max)}%',
                  null
                ),
                src: $src,
                unicode-range: $unicodeRange,
              ));
            }
          }
        }
      }
    }
  }
}

@mixin faces(
  $metadata: $metadata,
  $directory: $directory,
  $family: $family,
  $display: $display,
  $displayVar: $displayVar,
  $formats: $formats,
  $subsets: $subsets,
  $weights: $weights,
  $styles: $styles,
  $axes: $axes
) {
  @include generator(
      $metadata: $metadata,
      $directory: $directory,
      $family: $family,
      $display: $display,
      $displayVar: $displayVar,
      $formats: $formats,
      $subsets: $subsets,
      $weights: $weights,
      $styles: $styles,
      $axes: $axes
    )
    using ($props) {
    /* #{map.get($props, variant)} */
    @font-face {
      font-family: map.get($props, font-family);
      font-style: map.get($props, font-style);
      font-display: map.get($props, font-display);
      font-weight: map.get($props, font-weight);
      font-stretch: map.get($props, font-stretch);
      unicode-range: map.get($props, unicode-range);
      src: map.get($props, src);
    }
  }
}
`;
