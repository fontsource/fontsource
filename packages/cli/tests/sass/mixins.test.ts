import { compileString } from 'sass';
import { describe, expect, it } from 'vitest';

import { sassMetadata } from '../../src/sass/metadata';
import { sassMixins } from '../../src/sass/mixins';
import type { Metadata } from '../../src/types';
import mockSassMetadata from './fixtures/sass-metadata.json';

const compileSass = (metadata: string) => {
	const metadataMap = `(${metadata
		.replaceAll('$', '')
		.replaceAll(' !default;', ',')})`;

	return compileString(
		sassMixins
			.replace("@use 'metadata';", '')
			.replace('meta.module-variables(metadata) !default', metadataMap) +
			'@include faces()'
	);
};

describe('sass mixins', () => {
	it('should compile sass for non unicode range font successfully', async () => {
		expect(
			compileSass(
				sassMetadata(
					mockSassMetadata.carlito.metadata as Metadata,
					mockSassMetadata.carlito.unicode,
					false
				)
			)
		).toMatchSnapshot();
	});

	it('should compile sass for numeric subset font successfully', async () => {
		expect(
			compileSass(
				sassMetadata(
					mockSassMetadata.notoSansJp.metadata as Metadata,
					mockSassMetadata.notoSansJp.unicode,
					false
				)
			)
		).toMatchSnapshot();
	});

	it('should compile sass for variable font successfully', async () => {
		expect(
			compileSass(
				sassMetadata(
					mockSassMetadata.recursive.metadata as Metadata,
					mockSassMetadata.recursive.unicode,
					true
				)
			)
		).toMatchSnapshot();
	});
});
