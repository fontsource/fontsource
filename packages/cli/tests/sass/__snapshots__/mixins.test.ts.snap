// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sass mixins > should compile sass for non unicode range font successfully 1`] = `
{
  "css": "/* carlito-latin-400-normal */
@font-face {
  font-family: \\"Carlito\\";
  font-style: normal;
  font-display: var(--fontsource-display, swap);
  font-weight: 400;
  src: url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff2\\") format(\\"woff2\\"), url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff\\") format(\\"woff\\");
}",
  "loadedUrls": [],
}
`;

exports[`sass mixins > should compile sass for numeric subset font successfully 1`] = `
{
  "css": "/* carlito-latin-400-normal */
@font-face {
  font-family: \\"Carlito\\";
  font-style: normal;
  font-display: var(--fontsource-display, swap);
  font-weight: 400;
  src: url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff2\\") format(\\"woff2\\"), url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff\\") format(\\"woff\\");
}",
  "loadedUrls": [],
}
`;

exports[`sass mixins > should compile sass for variable font successfully 1`] = `
{
  "css": "/* carlito-latin-400-normal */
@font-face {
  font-family: \\"Carlito\\";
  font-style: normal;
  font-display: var(--fontsource-display, swap);
  font-weight: 400;
  src: url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff2\\") format(\\"woff2\\"), url(\\"~@fontsource/carlito/files/carlito-latin-400-normal.woff\\") format(\\"woff\\");
}",
  "loadedUrls": [],
}
`;
